common/OpTestSystem.py

```diff
@@ -222,91 +385,182 @@ def detect_target(self, target_state, reboot):
               return OpSystemState.UNKNOWN
           else:
             return OpSystemState.UNKNOWN
-        elif (r == 4) or (r == 5):
+        elif (r == 5) or (r == 6):
+          self.detect_counter += 1
           return OpSystemState.UNKNOWN
 
     def check_kernel(self):
-        console = self.console.get_console()
-        console.sendline("cat /proc/version | cut -d ' ' -f 3 | grep openpower")
-        console.expect("\n")
-        console.expect([".*#"])
-        console.sendline("echo $?")
-        console.expect("\n")
-        console.expect([".*#"])
-        echo_output = console.after
-        try:
-          echo_rc = int(echo_output.splitlines()[0])
-        except Exception as e:
-          # most likely cause is running while booting unknowlingly
-          return OpSystemState.UNKNOWN
-        if (echo_rc == 0):
-          return OpSystemState.PETITBOOT_SHELL
-        else:
-          return OpSystemState.OS
+        self.block_setup_term = 0 # unblock to allow setup_term during get_console
+        self.console.enable_setup_term_quiet()
+        sys_console = self.console.get_console()
+        self.console.disable_setup_term_quiet()
+        sys_console.sendline()
+        rc = sys_console.expect(["x=exit", "Petitboot", ".*#", ".*\$", "login:", pexpect.TIMEOUT, pexpect.EOF], timeout=5)
+        if rc in [0,1,5,6]:
+          return OpSystemState.UNKNOWN # we really should not have arrived in here and not much we can do
+        sys_console.sendline("cat /proc/version | cut -d ' ' -f 3 | grep %s; echo $?" % (self.openpower))
+        time.sleep(0.2)
+        rc = sys_console.expect([self.expect_prompt, pexpect.TIMEOUT, pexpect.EOF], timeout=1)
+        if rc == 0:
+          echo_output = sys_console.before
+          try:
+            echo_rc = int(echo_output.splitlines()[-1])
+          except Exception as e:
+            # most likely cause is running while booting unknowlingly
+            return OpSystemState.UNKNOWN
+          if (echo_rc == 0):
+            self.previous_state = OpSystemState.PETITBOOT_SHELL
+            return OpSystemState.PETITBOOT_SHELL
+          elif echo_rc == 1:
+            self.previous_state = OpSystemState.OS
+            return OpSystemState.OS
+          else:
+            return OpSystemState.UNKNOWN
+        else: # TIMEOUT EOF from cat
+            return OpSystemState.UNKNOWN
 
-    def wait_for_it(self, my_string=None, refresh=1, buffer_kicker=1, loop_max=50, threshold=1, timeout=10):
-        console = self.console.get_console()
+    def wait_for_it(self, **kwargs):
```

Comment:
I re-tested this latest code on a P9 system, and i see it is failing to reach petitboot state(3) as it is missing the string. This we need to really improve or atleast for now block the re-connect attempts as long as the console is stable we no need to worry for re-connecting. You can keep this new variant logic as optional, and use old method as a stable variant and let callers use this stable variant instead of optional one.

Reply:
> So it is in a loop forever even though system actually reached petitboot.

Yes, I found P9 system console output that gets to pexpect buffer can be very slow and so anytime it can miss catching anything, so I think the last_try helps catch and also that is what the logic of catching login is unexpected when trying to reach Petitboot and vice-versa if catching Petitboot if trying to go to login.  I think its the best case and it can never compensate for not getting data (either from slow console or from drops of connection like IPMI SOL).

I'll revisit and see if we can tune the non-IPMI cases a little better.

Reply:
> failing to reach petitboot state(3)

I also found on OpenBMC that if I tried manually through the port 2200 console it would not properly stop at the Petitboot if bootdev was set to stop, it would always go to OS.  Maybe check your system and see if this is happening.  If the system was set to stop at Petitboot then the buffer being slow is sync'd up and wait_for_it catches.

Reply:
My system is a boston system, which has all this functionality working, but with this new logic it may be failed to find petitboot string.
```
[   90.407191272,5] INIT: 64-bit LE kernel discovered
[   90.409329218,5] OCC: All Chip Rdy after 0 ms
[   90.438891161,5] INIT: Starting kernel at 0x20011000, fdt at 0x30a97f80 364786 bytes
[   92.111606245,5] OPAL: Switch to little-endian OS
ipmitool -H $BMCIP -I lanplus -U ADMIN -P ADMIN sol deactivate
#IPMI SOL CONNECT
[SOL Session operational.  Use ~? for help]
~. [terminated ipmitool]
ipmitool -H $BMCIP -I lanplus -U ADMIN -P ADMIN sol deactivate
#IPMI SOL CONNECT
[SOL Session operational.  Use ~? for help]
~. [terminated ipmitool]
ipmitool -H $BMCIP -I lanplus -U ADMIN -P ADMIN sol deactivate
#IPMI SOL CONNECT
[SOL Session operational.  Use ~? for help]
~. [terminated ipmitool]
ipmitool -H $BMCIP -I lanplus -U ADMIN -P ADMIN sol deactivate
#IPMI SOL CONNECT
[SOL Session operational.  Use ~? for help]
~. [terminated ipmitool]
ipmitool -H $BMCIP -I lanplus -U ADMIN -P ADMIN sol deactivate
#IPMI SOL CONNECT
[SOL Session operational.  Use ~? for help]
~. [terminated ipmitool]
ipmitool -H $BMCIP -I lanplus -U ADMIN -P ADMIN sol deactivate
#IPMI SOL CONNECT
[SOL Session operational.  Use ~? for help]
~. [terminated ipmitool]
```

After that i tried checking what is the state of the system.
it is in petitboot shell.
```
[SOL Session operational.  Use ~? for help]

/ # 
```

Reply:
> it is in petitboot shell.

If you can give me the system I can debug to check if its something new or if its the same cases I have seen.